modules/local/validatemodel/resources/usr/bin/validate_model.R

```diff
@@ -0,0 +1,434 @@
+
+        # Populate contrasts_list for later model validation
+        contrasts_list[[CONTRAST$id]] <- list(
+            "variable" = variable,
+            "contrast" = levels,
+            "blocking_factors" = CONTRAST$blocking_factors
+        )
+
+        # Gather contrasts per variable into temporary storage
+        if (variable %in% names(temp_var)) {
+            temp_var[[variable]] <- unique(c(temp_var[[variable]], levels))
+        } else {
+            temp_var[[variable]] <- levels
+        }
+
+        # Gather blocking factors
+        if (!is.null(CONTRAST$blocking_factors)) {
+            blocking_vars <- unique(c(blocking_vars, CONTRAST$blocking_factors))
+        }
+    }
+
+    # Consolidate gathered contrasts into `var`
+    for (variable in names(temp_var)) {
+        if (variable %in% names(var)) {
+            var[[variable]] <- unique(c(var[[variable]], temp_var[[variable]]))
+        } else {
+            var[[variable]] <- temp_var[[variable]]
+        }
+    }
+
+    # Add blocking variables to `var`
+    var[[ "blocking_factors" ]] <- blocking_vars
+
+    # Return the populated lists
+    return(list(
+        contrasts_list = contrasts_list,
+        var = var
+    ))
+}
+
+## Process models
+models_lists   <- process_models(models)
+### Extract models contrasts
+contrasts_list <- models_lists[[ "contrasts_list" ]]
+### Extract variables and levels
+var            <- models_lists[[ "var" ]]
+
+## Print explicit messages
+for (INDEX in 1:length(var)) {
+    cat("\033[32mDetected '", names(var)[INDEX], "' variable with ", paste(var[[INDEX]], collapse = " "), " levels.\033[0m \n", sep = "")
+}
+
+# VALIDATE SAMPLESHEET BASED ON YML FILE ----------------------------
+## Define function to compare YML and samplesheet
+validate_model <- function(sample_column, variables, samplesheet) { # sample_column: string; variables: list; samplesheet: df
+
+    #######################################################
+    ##
+    ## Function that takes a list of variables (and expected values) from the models.yml file that must be present in the sample sheet.
+    ##
+    ## The function evaluates that:
+    ## * Every variable is present in the sample sheet.
+    ## * Every variable has valid name
+    ## * There are no NAs
+    ## * All expected values are present
+    ## * All samples contains a valid level.
+    ##
+    ## Finally, the function returns
+    ## * An error vector containing all errors found.
+    ## * A warnings vector
+    ##
+    ## If the error vector contains values, the script will end with a non-zero status.
+    ##
+    #######################################################
+
+    ## Initialize errors vector
+    errors <- c()
+
+    ## Initialize warning errors
+    warnings <- c()
+
+    # Initialize vector to report continuos variables
+    continuous <- c()
+
+    ## Do not allow special characters
+    undesired_chars <- "[^a-zA-Z0-9_.]"
```

Comment:
Maybe precompile the regex to make it more efficient. I imagine this should work, but let me know if it doesn't.
undesired_chars_regex <- regex("[^a-zA-Z0-9_.]")